### ISSUE TITLE
Generated code in autoload_real.php should be namespaced

### ISSUE THREAD

**Issue by Firehed** [composer/composer#5231]
While collisions are effectively impossible due to the random (hashed?) class and function names, having the class and function in the global namespace is slightly annoying when doing work that involves `get_defined_functions` or `get_declared_classes` if any filtering of the resulting array is required, especially when using the namespace support in the Reflection APIs.

Unless there's a reason not to do this (which I'd be very interested in hearing), I'm more than happy to do the work and send a PR. I expect the file output would change from, roughly:
```php
<?php

// autoload_real.php @generated by Composer

class ComposerAutoloaderInit6c610ed51b1f80131597f878239bc803
{
   ...
}

function composerRequire6c610ed51b1f80131597f878239bc803($fileIdentifier, $file)
{
   ...
}
```

to
```php
<?php

// autoload_real.php @generated by Composer
namespace Composer\Autoloader;
class Init6c610ed51b1f80131597f878239bc803
{
    ...
}

function Require6c610ed51b1f80131597f878239bc803($fileIdentifier, $file)
{
    ...
}
```
and the related call sites would be updated.

The only question I have is whether `Composer\Autoload` (that ClassLoader uses) or `Composer\Autoloader` (based on the generated class name) would be preferable, or something else entirely.

**Comment by alcohol** [comment]
Well, the `Composer\Autoload` namespace exists already (see `src/Composer/Autoload`). Perhaps something that makes it easier to separate from the existing namespaces, `Composer\Generated` or `Composer\Vendor\Autoload`?

**Comment by Seldaek** [comment]
I *think* there is nothing speaking against it no. Sounds harmless enough but then again you never know what you're gonna break when changing such things. Since we allow setting the autoload suffix then people might rely on the class name for some reason. I don't see your use case as really warranting a BC break.

If anything you could also use https://getcomposer.org/doc/06-config.md#autoloader-suffix to set the suffix to a known value so that it gets easier to filter out?

**Comment by Firehed** [comment]
Unfortunately I can't since I'm building a standalone library, and the project root could override that value. There's really no great what that I've found to get at the whole project's metadata from the loaded Composer classes without manually re-parsing the json files (which is really my root issue). In any case the difficulty is the same in filtering by prefix, namespace, or suffux. What I hope to address is not needing to handle more than one, since I already need to filter out `Composer\Autoload\ClassLoader` so I'm *already* looking for the `Composer\` namespace (it's not a show-stopper by any means)

But I'm also keenly aware that the unlikely BC break this introduces would somewhat ironically most likely cause issues with my own use case ;) If there are future versions with BC breaks planned, would you consider making (or merging) this change?

**Comment by Seldaek** [comment]
Well.. even if we make a breaking release, the amount of breakage should only be as great as needed ideally. Breaking things for the sake of breaking things isn't great :) And in this case really I am not sure the benefit is that great. You have one prefix or two to filter, doesn't seem like it's worth inflicting pain on anyone else :)

**Comment by Firehed** [comment]
Of course, it's totally reasonable to avoid adding any unnecessary BC breaks, even during a release that allows for them. I wasn't trying to suggest doing it just to do it, even if my specific use-case may come across as that way.

As an end-user, I would still expect a project using namespaces to not put code in the global namespace, even if that code is automatically generated. Composer is obviously pretty special both within the PHP community and in its specific needs, but I'd still consider the current behavior both unexpected and undesirable.

But ultimately it's your call and I respect the decision either way.

**Comment by Seldaek** [comment]
I tend to agree in general, which is why it's prefixed with `composer*` (and hence namespaced in a way). I guess we could do this in 2.0 and if anyone complains before we release a stable 2.0 we can always revert.
